modules/ipp/scanner.go

```diff
@@ -126,86 +172,394 @@ func (scanner *Scanner) GetPort() uint {
+	b := bufferFromBody(res, scanner)
+	res.BodyText = b.String()
+	if len(res.BodyText) > 0 {
+		m := sha256.New()
+		m.Write(b.Bytes())
+		res.BodySHA256 = m.Sum(nil)
+	}
 }
 
-// FIXME: Change to instead return (*ScanResults, *zgrab2.ScanError)
-// TODO: Doesn't support TLS at all right now
-func (scanner *Scanner) grab(target zgrab2.ScanTarget) (int8, int8, *zgrab2.ScanError) {
-	//FIXME: This is not where this hostname assignment logic should live
-	host := target.Domain
-	if host == "" {
-		host = target.IP.String()
+// TODO: Make sure this it isn't too slow to use this instead of copy-pasting everywhere necessary
+func bufferFromBody(res *http.Response, scanner *Scanner) *bytes.Buffer {
+	b := new(bytes.Buffer)
+	maxReadLen := int64(scanner.config.MaxSize) * 1024
+	readLen := maxReadLen
+	if res.ContentLength >= 0 && res.ContentLength < maxReadLen {
+		readLen = res.ContentLength
+	}
+	io.CopyN(b, res.Body, readLen)
+	res.Body.Close()
+	res.Body = ioutil.NopCloser(b)
+	return b
+}
+
+// FIXME: add proper error handling to this
+// TODO: Support reading from multiple instances of the same attribute in a response
+func readAttributeFromBody(attrString string, body *[]byte) ([][]byte, error) {
+	attr := []byte(attrString)
+	interims := bytes.Split(*body, attr)
+	if len(interims) > 1 {
+		valueTag := interims[0][len(interims[0])-3]
+		var vals [][]byte
+		buf := bytes.NewBuffer(interims[1])
+		for tag, nameLength := valueTag, int16(0); tag == valueTag && nameLength == 0; {
+			var length int16
+			if err := binary.Read(buf, binary.BigEndian, &length); err != nil {
+				//Couldn't read length of content for some reason
+				//return nil, err
+				continue
+			}
+			val := make([]byte, length)
+			if err := binary.Read(buf, binary.BigEndian, &val); err != nil {
+				//Content not read without issue.
+				//vals = append(vals, val)
+				//return vals, err
+				continue
+			}
+			//return &val, nil
+			vals = append(vals, val)
+			if err := binary.Read(buf, binary.BigEndian, &tag); err != nil {
+				//Couldn't read next valueTag
+				//break
+				continue
+			}
+			if err := binary.Read(buf, binary.BigEndian, &nameLength); err != nil {
+				//Couldn't read next nameLength
+				//break
+				// TODO: Conclude correct error behavior for all of these blocks
+				continue
+			}
+		}
+		return vals, nil // FIXME: This should be whatever error occurred during a break case
 	}
-	//FIXME: ?Should just use endpoint "/", since we get the same response as "/ipp" on CUPS??
-	uri := getIPPURL(scanner.config.IPPSecure, host, uint16(scanner.config.BaseFlags.Port), "/ipp")
-	b := getPrinterAttributesRequest(uri)
+	//The attribute was not present
+	return nil, errors.New("Attribute \"" + attrString + "\" not present.")
+}
 
-	resp, err := http.Post(uri, ContentType, &b)
+func versionNotSupported(resp *http.Response, scanner *Scanner) bool {
+	if resp != nil && resp.Body != nil {
+		buf := bufferFromBody(resp, scanner)
```

Comment:
Am I missing something, or won't this wipe out `resp.Body` -- so any subsequent calls (like `storeBody()` on 299) will get nothing back?

The way I read it, it seems like `storeBody()` should be called first, then `versionNotSupported` called on that -- but I may be missing something.

Reply:
I've been struggling with how to handle that all correctly. `bufferFromBody()` returns a buffer filled with the contents of the body, but should assign a new io.ReadCloser with the original contents to `resp.Body`*. My understanding was that subsequent calls reading from `resp.Body` (like `storeBody()`) should find a non-empty ReadCloser and be fine. But I've been having some issues with this. Basically `versionNotSupported` was exhibiting incorrect behavior when executed after `storeBody()`. So something's definitely off.

*(Technically it's only the first `maxReadLen` KB of `resp.Body` because that's all we'll read anyway).

Reply:
Could you call `storeBody()` once at the start, then always use `response.BodyText`? I may be missing some problem with that, but that was what I was expecting to see.